### PR TITLE
Add support for @@iterator

### PR DESCRIPTION
**Pull request by rajzshkr**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes https://github.com/babel/babel/issues/7053
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR         | 
| Any Dependency Changes?  |
| License                  | MIT

### CODE REVIEW

File: packages/babylon/src/tokenizer/index.js [hunk 1]
```diff
@@ -1231,6 +1231,9 @@ export default class Tokenizer extends LocationParser {
       const ch = this.fullCharCodeAtPos();
       if (isIdentifierChar(ch)) {
         this.state.pos += ch <= 0xffff ? 1 : 2;
+      } else if (this.state.isIterator) {
+        this.state.pos += 2;
+        delete this.state.isIterator;
```
Comment (nicolo-ribaudo) [inline packages/babylon/src/tokenizer/index.js]: Can you use `this.state.isIterator` here?

**Comment by hzoo** [comment]
@nicolo-ribaudo it's a flow syntax not related to JS, it would error with `let @@iterator = 1`? Also funny since the link to flow says TODO switch with Symbol.iterator?

File: packages/babylon/src/tokenizer/index.js (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babylon/src/tokenizer/index.js]: I meant `this.state.isIterator = false` lol

**Comment by nicolo-ribaudo** [comment]
> Also funny since the link to flow says TODO switch with Symbol.iterator?

We should ask to the flow team if they plan to remove `@@iterator` in the near future.

**Comment by hzoo** [comment]
Should ping the flow team on all the flow prs anyway 😛

**Comment by hzoo** [comment]
@bcherny

**Comment by rajzshkr** [comment]
Also we should not allow any other identifier `iterator` and `asyncIterator`. I'll update the PR

**Comment by rajzshkr** [comment]
> Should the @@iterator identifier be allowed everywhere? e.g. let @@iterator = 2;

Flow parse the above syntax, but in V8 throws an error.

**Comment by nicolo-ribaudo** [comment]
> Flow parse the above syntax

I checked and flow paeses that syntax *sometimes* (e.g. it accepts `let @@iterator = number` but not `type @@iterator = number`.
I think it's ok for us to allow it only in class/object/interface property types.

**Comment by nicolo-ribaudo** [comment]
I have thought about this a little, and I think that we should allow `@@iterator` **only** in class/object/interface.
The flow parser is more lenient and accepts things like
```js
function @@iterator() {}
@@iterator();
```
but I think that we shouldn't parse it since it will generate invalid JavaScript output.

So, instead of throwing if `@@iterator` is after var/let/type we should throw if it is *not* inside an object/interface/class property name.

**Comment by existentialism** [comment]
@nicolo-ribaudo agreed

**Comment by loganfsmyth** [comment]
@danez Would you like to look this over since you're familiar with the parser? It seems fine to me, but I'm also not an expert on it.

**Comment by nicolo-ribaudo** [comment]
> Flow-strip-types probably needs to remove the @@iterator and @@asyncIterator, as the resulting code is otherwise invalid?

Currently Flow only removes them from types.

**Comment by xtuc** [comment]
Looks good to me. As mentioned by @danez, `babel-generator` will probably need to be updated in order to emit this syntax and could you please add a test for it?

It could probably also impact TypeScript? @hzoo could we ping them? I don't think that could break anything but maybe require a transformations (I'm not a TS expert).

**Comment by rajzshkr** [comment]
@nicolo-ribaudo updated the test cases.

**Comment by rajzshkr** [comment]
@danez I added test cases, to make sure iterator is removed from types in Flow-strip-types. https://github.com/babel/babel/pull/7058/commits/36a07317e8fddf4b7abe420b79d054310917fa8c, Do you want to add test case for babel-generator, since it is already removed by Flow-strip-types

**Comment by nicolo-ribaudo** [comment]
Yes, since someone might use Babel for a codemod and thus not remove Flow types.

**Comment by rajzshkr** [comment]
@nicolo-ribaudo @danez added test case for generator and Flow-strip-types plugin

File: packages/babel-plugin-transform-flow-strip-types/test/fixtures/strip-types/strip-iterator/input.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+declare class A {
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-flow-strip-types/test/fixtures/strip-types/strip-iterator/input.js]: It looks like the output for this file is missing?

Comment (existentialism) [reply packages/babel-plugin-transform-flow-strip-types/test/fixtures/strip-types/strip-iterator/input.js]: That's expected (i.e. there is no output)

Comment (rajzshkr) [reply packages/babel-plugin-transform-flow-strip-types/test/fixtures/strip-types/strip-iterator/input.js]: @existentialism Thanks!!

**Comment by existentialism** [comment]
@rajzshkr thanks!

**Pull request merged**